File: sdk/java/src/main/java/com/infergo/VectorDB.java
package com.infergo;

/**
 * Persistent HNSW vector database.
 *
 * <pre>{@code
 * try (VectorDB db = new VectorDB(384, 16, 200)) {
 *     db.insert(1, embedding, "{\"title\": \"doc1\"}");
 *     VectorDB.SearchResult[] results = db.search(queryVec, 5, 100, null);
 *     for (VectorDB.SearchResult r : results) {
 *         System.out.printf("id=%d dist=%.4f%n", r.id, r.distance);
 *     }
 * }
 * }</pre>
 */
public class VectorDB implements AutoCloseable {

    private long handle;

    /**
     * A single search result: ID + distance.
     */
    public static class SearchResult {
        public final long id;
        public final float distance;

        public SearchResult(long id, float distance) {
            this.id = id;
            this.distance = distance;
        }

        @Override
        public String toString() {
            return "SearchResult{id=" + id + ", distance=" + distance + "}";
        }
    }

    /**
     * Create a persistent vector database.
     *
     * @param dim            vector dimension (must match embedding model output)
     * @param M              max connections per HNSW node (16 = good default)
     * @param efConstruction search width during build (200 = good default)
     * @throws InfergoException if creation fails
     */
    public VectorDB(int dim, int M, int efConstruction) {
        this.handle = InfergoNative.vectordbCreate(dim, M, efConstruction);
        if (this.handle == 0) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException("Failed to create VectorDB: " + (err != null ? err : "unknown error"));
        }
    }

    /**
     * Insert a vector with metadata.
     *
     * @param id       unique vector ID
     * @param vec      float array of length dim
     * @param metadata JSON metadata string, or null
     * @throws InfergoException on failure
     */
    public void insert(long id, float[] vec, String metadata) {
        checkOpen();
        int rc = InfergoNative.vectordbInsert(handle, id, vec, metadata);
        if (rc != 0) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException(rc, "VectorDB insert failed: " + (err != null ? err : "unknown error"));
        }
    }

    /**
     * Insert a vector without metadata.
     */
    public void insert(long id, float[] vec) {
        insert(id, vec, null);
    }

    /**
     * Delete a vector by ID.
     *
     * @param id vector ID to delete
     * @throws InfergoException on failure
     */
    public void delete(long id) {
        checkOpen();
        int rc = InfergoNative.vectordbDelete(handle, id);
        if (rc != 0) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException(rc, "VectorDB delete failed: " + (err != null ? err : "unknown error"));
        }
    }

    /**
     * Search for k nearest neighbors.
     *
     * @param query          query vector
     * @param k              number of results
     * @param efSearch       search width (higher = more accurate, slower)
     * @param metadataFilter metadata filter expression, or null
     * @return array of search results sorted by distance
     * @throws InfergoException on failure
     */
    public SearchResult[] search(float[] query, int k, int efSearch, String metadataFilter) {
        checkOpen();
        long[] ids = InfergoNative.vectordbSearchIds(handle, query, k, efSearch,
                                                     metadataFilter, k);
        float[] distances = InfergoNative.vectordbSearchDistances(handle, query, k, efSearch,
                                                                  metadataFilter, k);
        if (ids == null || distances == null) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException("VectorDB search failed: " + (err != null ? err : "unknown error"));
        }

        int n = Math.min(ids.length, distances.length);
        SearchResult[] results = new SearchResult[n];
        for (int i = 0; i < n; i++) {
            results[i] = new SearchResult(ids[i], distances[i]);
        }
        return results;
    }

    /**
     * Search without metadata filter.
     */
    public SearchResult[] search(float[] query, int k, int efSearch) {
        return search(query, k, efSearch, null);
    }

    /**
     * Save the database to a file.
     *
     * @param path file path
     * @throws InfergoException on failure
     */
    public void save(String path) {
        checkOpen();
        int rc = InfergoNative.vectordbSave(handle, path);
        if (rc != 0) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException(rc, "VectorDB save failed: " + (err != null ? err : "unknown error"));
        }
    }

    /**
     * Load the database from a file.
     *
     * @param path file path
     * @throws InfergoException on failure
     */
    public void load(String path) {
        checkOpen();
        int rc = InfergoNative.vectordbLoad(handle, path);
        if (rc != 0) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException(rc, "VectorDB load failed: " + (err != null ? err : "unknown error"));
        }
    }

    /**
     * Returns the number of vectors in the database.
     */
    public int size() {
        checkOpen();
        return InfergoNative.vectordbSize(handle);
    }

    /**
     * Returns the native handle. For advanced use only.
     */
    public long getHandle() {
        checkOpen();
        return handle;
    }

    @Override
    public void close() {
        if (handle != 0) {
            InfergoNative.vectordbFree(handle);
            handle = 0;
        }
    }

    private void checkOpen() {
        if (handle == 0) {
            throw new IllegalStateException("VectorDB is closed");
        }
    }
}

File: sdk/dotnet/src/Infergo/VectorDb.cs
namespace Infergo;

/// <summary>
/// Search result from the vector database.
/// </summary>
public readonly record struct VectorSearchResult(long Id, float Distance);

/// <summary>
/// Managed wrapper for the infergo HNSW vector database.
/// </summary>
public sealed class VectorDb : IDisposable
{
    private IntPtr _handle;
    private bool _disposed;
    private readonly int _dim;

    /// <summary>
    /// Create a new in-memory vector database.
    /// </summary>
    /// <param name="dim">Vector dimension (must match embedding model output).</param>
    /// <param name="m">HNSW M parameter — max connections per node (16 is a good default).</param>
    /// <param name="efConstruction">HNSW ef_construction — search width during build (200 is a good default).</param>
    public VectorDb(int dim, int m = 16, int efConstruction = 200)
    {
        if (dim <= 0) throw new ArgumentOutOfRangeException(nameof(dim));

        _dim = dim;
        _handle = Native.infer_vectordb_create(dim, m, efConstruction);
        if (_handle == IntPtr.Zero)
            throw InfergoException.FromNative("Failed to create vector database");
    }

    /// <summary>
    /// Vector dimension.
    /// </summary>
    public int Dimension => _dim;

    /// <summary>
    /// Number of vectors in the database.
    /// </summary>
    public int Count
    {
        get
        {
            ObjectDisposedException.ThrowIf(_disposed, this);
            return Native.infer_vectordb_size(_handle);
        }
    }

    /// <summary>
    /// Insert a vector with an ID and optional metadata.
    /// </summary>
    /// <param name="id">Unique document ID.</param>
    /// <param name="vector">Float array of length <see cref="Dimension"/>.</param>
    /// <param name="metadata">Optional JSON metadata string.</param>
    public void Insert(long id, float[] vector, string? metadata = null)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(vector);
        if (vector.Length != _dim)
            throw new ArgumentException($"Vector length {vector.Length} does not match dimension {_dim}.", nameof(vector));

        int rc = Native.infer_vectordb_insert(_handle, id, vector, metadata);
        if (rc != 0)
            throw InfergoException.FromNative($"VectorDB insert failed for id={id}", rc);
    }

    /// <summary>
    /// Delete a vector by ID.
    /// </summary>
    public void Delete(long id)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        int rc = Native.infer_vectordb_delete(_handle, id);
        if (rc != 0)
            throw InfergoException.FromNative($"VectorDB delete failed for id={id}", rc);
    }

    /// <summary>
    /// Update a vector and its metadata.
    /// </summary>
    public void Update(long id, float[] vector, string? metadata = null)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(vector);
        if (vector.Length != _dim)
            throw new ArgumentException($"Vector length {vector.Length} does not match dimension {_dim}.", nameof(vector));

        int rc = Native.infer_vectordb_update(_handle, id, vector, metadata);
        if (rc != 0)
            throw InfergoException.FromNative($"VectorDB update failed for id={id}", rc);
    }

    /// <summary>
    /// Search for the k nearest neighbors to a query vector.
    /// </summary>
    /// <param name="query">Query vector of length <see cref="Dimension"/>.</param>
    /// <param name="k">Number of results to return.</param>
    /// <param name="efSearch">Search width (higher = more accurate but slower; 50-200 typical).</param>
    /// <param name="metadataFilter">Optional metadata filter expression.</param>
    /// <returns>Array of search results sorted by distance (ascending).</returns>
    public VectorSearchResult[] Search(float[] query, int k = 10, int efSearch = 100, string? metadataFilter = null)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(query);
        if (query.Length != _dim)
            throw new ArgumentException($"Query length {query.Length} does not match dimension {_dim}.", nameof(query));
        if (k <= 0) throw new ArgumentOutOfRangeException(nameof(k));

        long[] ids = new long[k];
        float[] distances = new float[k];

        int found = Native.infer_vectordb_search(_handle, query, k, efSearch, metadataFilter, ids, distances, k);
        if (found < 0)
            throw InfergoException.FromNative("VectorDB search failed");

        var results = new VectorSearchResult[found];
        for (int i = 0; i < found; i++)
            results[i] = new VectorSearchResult(ids[i], distances[i]);
        return results;
    }

    /// <summary>
    /// Save the database to a file.
    /// </summary>
    public void Save(string path)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(path);

        int rc = Native.infer_vectordb_save(_handle, path);
        if (rc != 0)
            throw InfergoException.FromNative($"VectorDB save failed: {path}", rc);
    }

    /// <summary>
    /// Load the database from a file.
    /// </summary>
    public void Load(string path)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(path);

        int rc = Native.infer_vectordb_load(_handle, path);
        if (rc != 0)
            throw InfergoException.FromNative($"VectorDB load failed: {path}", rc);
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;

        if (_handle != IntPtr.Zero)
        {
            Native.infer_vectordb_free(_handle);
            _handle = IntPtr.Zero;
        }
    }
}
